lint when calling the blanket `Into` impl from a `From` impl

y21 opened the pull request:
Closes #11150
```
warning: function cannot return without recursing
  --> x.rs:9:9
   |
9  | /         fn from(value: f32) -> Self {
10 | |             value.into()
11 | |         }
   | |_________^
   |
note: recursive call site
  --> x.rs:10:13
   |
10 |             value.into()
   |             ^^^^^^^^^^^^
```

I'm also thinking that we can probably generalize this lint to #11032 at some point (instead of hardcoding a bunch of impls), like how rustc's `unconditional_recursion` works, at least up to one indirect call, but this still seems useful for now :)

I've also noticed that we use `fn_def_id` in a bunch of lints and then try to get the node args of the call in a separate step, so I made a helper function that does both in one. I intend to refactor a bunch of uses of `fn_def_id` to use this later

I can add more test cases, but this is already using much of the same logic that exists for the other impls that this lint looks for (e.g. making sure that there are no conditional returns).

changelog: [`unconditional_recursion`]: emit a warning inside of `From::from` when unconditionally calling the blanket `.into()` impl

hamirmahal commented on clippy_lints/src/unconditional_recursion.rs:
```diff
@@ -19,11 +20,11 @@ use rustc_trait_selection::traits::error_reporting::suggestions::ReturnsVisitor;
     ///
     /// ### Why is this bad?
-    /// This is a hard to find infinite recursion which will crashing any code
+    /// This is a hard to find infinite recursion which will crash any code
```
```suggestion
    /// This is a hard to find infinite recursion that will crash any code.
```

Jarcho commented on clippy_lints/src/unconditional_recursion.rs:
```diff
@@ -394,6 +393,33 @@ impl UnconditionalRecursion {
+    // which calls back into our `From::from` again (`Into` is not specializable).
+    // rustc's unconditional_recursion already catches calling `From::from` directly
+    if let Some((fn_def_id, node_args)) = fn_def_id_with_node_args(cx, expr)
+        && let Some((s1, s2)) = node_args.iter().filter_map(ty::GenericArg::as_type).collect_tuple()
```
```suggestion
&& let [s1, s2] = node_args
&& let (Some(s1), Some(s2)) = (s1.as_type(), s2.as_type())
```

Jarcho reviewed (commented):
This should have some negative tests. Otherwise Looks good.

y21 commented:
added a few more tests I could think of in the last commit, if you have some more concrete ones I can add them. Also removed an unnecessary `get_trait_def_id` which could be using diagnostic items

Jarcho reviewed (approved):
@bors delegate

All looks good. r=me after squashing.

y21 commented:
squashed
@bors r=Jarcho

bors commented:
:pushpin: Commit 65defdb47482e26037ff9e5ca311522c7823372f has been approved by `Jarcho`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Jarcho 65defdb47482e26037ff9e5ca311522c7823372f -->
<!-- homu: {"type":"Approved","sha":"65defdb47482e26037ff9e5ca311522c7823372f","approver":"Jarcho","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 65defdb47482e26037ff9e5ca311522c7823372f with merge 73be4863f09255730e33bdd27f20154baad44203...
<!-- homu: {"type":"BuildStarted","head_sha":"65defdb47482e26037ff9e5ca311522c7823372f","merge_sha":"73be4863f09255730e33bdd27f20154baad44203"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/8268240222/job/22620663643), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/8268240214/job/22620677364), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/8268240224/job/22620995548)
Approved by: Jarcho
Pushing 73be4863f09255730e33bdd27f20154baad44203 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Jarcho","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8268240224/job/22620995548","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8268240222/job/22620663643","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8268240214/job/22620677364"},"merge_sha":"73be4863f09255730e33bdd27f20154baad44203"} -->
